### PR TITLE
install package with file: protocol as default if it exists in the filesystem and using fs.exists instead of using fs.existsSync

### LINKED ISSUE #605: Support local file dependencies without file protocol
**Do you want to request a _feature_ or report a _bug_?**
bug

**What is the current behavior?**
In a number of my projects I have a package.json with dependencies as such:

``` javascript
dependencies: {
    "babel-preset-something": "../ui-babel-preset-something",
}
```

`yarn` fails:

```
> yarn
yarn install v0.15.0
info No lockfile found.
[1/4] 🔍  Resolving packages...
error Couldn't find package "babel-preset-something" on the "npm" registry.
info Visit http://yarnpkg.com/en/docs/cli/install for documentation about this command.
```

**If the current behavior is a bug, please provide the steps to reproduce.**

``` shell
> cd yarn-test
> ls
> mkdir a b
> cd a
> ~/.yarn/bin/yarn.js init
yarn init v0.15.0
question name (a):
question version (1.0.0):
question description:
question entry point (index.js):
question git repository:
question author:
question license (MIT):
success Saved package.json
✨  Done in 1.97s.
> cd ../b
> ~/.yarn/bin/yarn.js init
yarn init v0.15.0
question name (b):
question version (1.0.0):
question description:
question entry point (index.js):
question git repository:
question author:
question license (MIT):
success Saved package.json
✨  Done in 1.88s.
> ~/.yarn/bin/yarn.js add ../a
yarn add v0.15.0
info No lockfile found.
[1/4] 🔍  Resolving packages...
error Couldn't find package "../a" on the "npm" registry.
info Visit http://yarnpkg.com/en/docs/cli/add for documentation about this command.
```

**What is the expected behavior?**

Expected behavior is that running `yarn` installs the package into node_modules.

**Please mention your node.js, yarn and operating system version.**
- node v6.2.0
- yarn 0.15.0
- OSX 10.11.3

### PR DESCRIPTION
**Pull request by voxsim**
This pull request should resolve #605 and it is related to https://github.com/yarnpkg/yarn/pull/2684.

In this case we use `fs.exists` instead of using `fs.existsSync`.

cheers,
voxsim

### CODE REVIEW

File: src/package-request.js [hunk 1]
```diff
@@ -136,6 +138,30 @@ export default class PackageRequest {
...
+    if (await fs.exists(path.join(this.config.cwd, pattern))) {
+      return Promise.resolve(`file:${pattern}`);
+    }
+
+    return Promise.resolve(pattern);
+  }
+
+  async normalize(pattern: string): any {
+    const {name, range, hasVersion} = PackageRequest.normalizePattern(pattern);
+    const new_range = await this.normalizeRange(range);
```
Comment (bestander) [inline src/package-request.js]: nit: do we use underscore for word separation anywhere in Yarn?
I think we should stick to the same camelCase.

File: src/package-request.js [hunk 2]
```diff
@@ -136,6 +138,30 @@ export default class PackageRequest {
...
+   * - contains : it will return because is something like '<protocol>:<something else' or an url
+   * - is a file it will prepend file:
+   * otherwise it will return
+   */
+  async normalizeRange(pattern: string): Promise<string> {
+    if (pattern.includes(':')) {
+      return Promise.resolve(pattern);
+    }
+
+    if (await fs.exists(path.join(this.config.cwd, pattern))) {
```
Comment (bestander) [inline src/package-request.js]: How often is this IO operation is used?
Can we infer if pattern looks like a file first before touching the disk if this happens often?

Comment (voxsim) [inline src/package-request.js]: After some over-thinking we can do something like this:

```
    if (pattern.includes(':') ||
      pattern.includes('@') ||
      PackageRequest.getExoticResolver(pattern)) {
      return Promise.resolve(pattern);
    }
```
this should give us enough confidence that the pattern may be a file if we go ahead in the method. Do you like it? Do you think is enough?

File: src/package-request.js (hunk 1 above)
Comment (voxsim) [inline src/package-request.js]: done, resolved ;)

**Comment by bestander** [comment]
Thanks for the fix, @voxsim and for doing a quick turnaround!

**Comment by kjbkjb** [comment]
greetings.

Very new to yarn. Loving it.

I believe this particular PR has broken local tgz file installation.

```
  "dependencies": {
    "bell": "^8.6.0",
    "boom": "^4.3.1",
    "code": "^4.0.0",
    "confidence": "^3.0.2",
    "config": "^1.12.0",
    "crumb": "^6.0.3",
    "dustjs-linkedin": "^2.7.2",
    "feature-toggles": "file:.modules/feature-toggles-3.0.3.tgz",
    "global-session": "file:.modules/global-session-3.0.0.tgz",
```

This mechanism worked in 0.22, but fails in 0.23

Now we are seeing:

```
yarn
yarn install v0.23.2
[1/4] 🔍  Resolving packages...
error An unexpected error occurred: "https://npm-proxy.fury.io/Up97EjxiNexkaesDEnU1/sportngin/.modules/sportngin-feature-toggles-3.0.3.tgz: Request failed \"404 Not Found\"".
info If you think this is a bug, please open a bug report with the information provided in "/Users/kevinboedigheimer/workspace/payments/yarn-error.log".
info Visit https://yarnpkg.com/en/docs/cli/install for documentation about this command.
```

**Comment by voxsim** [comment]
hey @kjbkjb, have you tried with another version g the same package? Because the issue it does not seem related, i mean really the serve doesn't have that version of the package. Btw if you think it is a real issue, maybe it's better to open an issue and link this pull request ;)

**Comment by kjbkjb** [comment]
I've tried building several fresh versions. if the version did not exist before and it was not listed in the lock file, version 0.22 would find the file in the local file system, but 0.23 tries to look up the module on npm and ignores the local file. This pr was dealing with file installation so I thought it looked suspicious. I'll file an issue.

**Pull request merged**